### PR TITLE
Make tot able to fallback to latest-build.txt for new builds.

### PR DESCRIPTION
**Pull request by rmmh**

### CODE REVIEW

**Comment by krzyzacy** [comment]
that's my initial idea, @spxtr might not like it. I'm fine with having it and move Jenkins stuff over until we switch to timestamp and we can retire tot.

But it will not help the jobs that are already migrated though since latestbuild is overwritten by bootstrap.

**Comment by rmmh** [comment]
The first commit will help with that.

**Comment by krzyzacy** [comment]
with some manual set calls?

**Comment by rmmh** [comment]
Yes. The current state is unfortunate, but we can recover from it relatively easily.

File: prow/cmd/tot/main.go [hunk 1]
```diff
@@ -79,13 +83,92 @@ func (s *store) save() error {
...
+}
+
+func (f fallbackHandler) get(b string) int {
+	url := fmt.Sprintf(f.template, b)
+
+	var body []byte
+
+	for i := 0; i < 10; i++ {
+		resp, err := http.Get(url)
+		if err == nil && resp.StatusCode == http.StatusOK {
```
Comment (krzyzacy) [inline prow/cmd/tot/main.go]: `defer resp.Body.Close()`?

File: prow/cmd/tot/main_test.go [hunk 2]
```diff
@@ -30,28 +34,114 @@ func expectEqual(t *testing.T, msg string, have interface{}, want interface{}) {
...
+	expectResponse(t, handler, req, "http peek", "2")
+
+	req, err = http.NewRequest("POST", "/vend/bar", strings.NewReader("40"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expectResponse(t, handler, req, "http post", "")
+
+	req, err = http.NewRequest("HEAD", "/vend/bar", nil)
```
Comment (krzyzacy) [inline prow/cmd/tot/main_test.go]: check err

File: prow/cmd/tot/main_test.go [hunk 3]
```diff
@@ -30,28 +34,114 @@ func expectEqual(t *testing.T, msg string, have interface{}, want interface{}) {
...
+
+	expectResponse(t, handler, req, "http vend", "40")
+}
+
+type mapHandler map[string]string
+
+func (h mapHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	fmt.Printf("what? %s", r.URL.Path)
```
Comment (krzyzacy) [inline prow/cmd/tot/main_test.go]: remove debug info?

**Comment by krzyzacy** [comment]
I'm fine with it, go persuade @spxtr 
xref https://github.com/kubernetes/test-infra/issues/2996

File: prow/cmd/tot/main.go [hunk 4]
```diff
@@ -35,12 +37,14 @@ var (
 	port        = flag.Int("port", 8888, "port to listen on")
 	logJson     = flag.Bool("log-json", false, "output log in JSON format")
 	storagePath = flag.String("storage", "tot.json", "where to store the results")
+	fallbackURI = flag.String("fallback", "", "HTTP endpoint to GET for missing build numbers -- template with %s")
```
Comment (kargakis) [inline prow/cmd/tot/main.go]: Are you missing something for the formatting of the help message?

Comment (krzyzacy) [inline prow/cmd/tot/main.go]: oh, escape % :-)

**Comment by spxtr** [comment]
One possibility is to have prow use a timestamp or UUID and have `bootstrap.py` do the build-number generation. The only downside is that prow will link to `gubernator/build/kubernetes-jenkins/pr-logs/pull/47099/pull-kubernetes-node-e2e/BIG-LONG-UUID/` and so on, so bootstrap would need to also upload a pointer from there to the actual location. @rmmh wdyt?

File: prow/cmd/tot/main.go (hunk 4 above)
Comment (spxtr) [inline prow/cmd/tot/main.go]: It is correct as is, but confusing and unidiomatic. I wouldn't even bother allowing templating, just hard-code the gubernator link. We're going to add this sort of thing to a config file in the future, and I'd rather avoid ad-hoc templates in arguments.

File: prow/cmd/tot/main.go [hunk 5]
```diff
@@ -90,6 +99,83 @@ func (s *store) vend(b string) int {
 	return n
 }
 
+func (s *store) peek(b string) int {
```
Comment (spxtr) [inline prow/cmd/tot/main.go]: This ignores fallback?

File: prow/cmd/tot/main.go [hunk 6]
```diff
@@ -90,6 +99,83 @@ func (s *store) vend(b string) int {
...
+		log.Infof("vending %s number %d to %s", b, n, r.RemoteAddr)
+		fmt.Fprintf(w, "%d", n)
+	case "HEAD":
+		n := s.peek(b)
+		log.Infof("peeking %s number %d to %s", b, n, r.RemoteAddr)
+		fmt.Fprintf(w, "%d", n)
+	case "POST":
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Error("unable to read body", err)
```
Comment (spxtr) [inline prow/cmd/tot/main.go]: `log.WithError(err).Error("Unable to read body.")`

File: prow/cmd/tot/main.go [hunk 7]
```diff
@@ -90,6 +99,83 @@ func (s *store) vend(b string) int {
...
+			return
+		}
+		n, err := strconv.Atoi(string(body))
+		if err != nil {
+			log.Error("unable to parse number", err)
```
Comment (spxtr) [inline prow/cmd/tot/main.go]: as above

File: prow/cmd/tot/main.go [hunk 8]
```diff
@@ -90,6 +99,83 @@ func (s *store) vend(b string) int {
...
+		if err == nil {
+			defer resp.Body.Close()
+			if resp.StatusCode == http.StatusOK {
+				body, err = ioutil.ReadAll(resp.Body)
+				if err == nil {
+					break
+				}
+			}
+		}
+		log.Error("fallback failed with", err)
```
Comment (spxtr) [inline prow/cmd/tot/main.go]: as above

File: prow/cmd/tot/main.go [hunk 9]
```diff
@@ -90,6 +99,83 @@ func (s *store) vend(b string) int {
...
+		time.Sleep(2)
+	}
+
+	n, err := strconv.Atoi(strings.TrimSpace(string(body)))
+
```
Comment (spxtr) [inline prow/cmd/tot/main.go]: remove this newline

File: prow/cmd/tot/main.go (hunk 5 above)
Comment (rmmh) [inline prow/cmd/tot/main.go]: Yes, since it's looking at the *current* state.

**Pull request merged**